Added 'gymname_contains' to gym filters.

tomballgithub opened the pull request:
<!--       PULL REQUESTS CREATED NOT USING ONE OF THE BELOW         --->
<!--      ISSUE TEMPLATES WILL BE CLOSED WITHOUT ANY RESPONSE       --->
<!--       PULL REQUESTS CREATED NOT USING ONE OF THE BELOW         --->
<!--      ISSUE TEMPLATES WILL BE CLOSED WITHOUT ANY RESPONSE       --->
<!--       PULL REQUESTS CREATED NOT USING ONE OF THE BELOW         --->
<!--      ISSUE TEMPLATES WILL BE CLOSED WITHOUT ANY RESPONSE       --->
<!--       PULL REQUESTS CREATED NOT USING ONE OF THE BELOW         --->
<!--      ISSUE TEMPLATES WILL BE CLOSED WITHOUT ANY RESPONSE       --->


<!--
 Please make all PRs to the dev branch. The dev branch will be
 periodically pulled into the master branch. This allows time for
 the changes and documentation to be tested before being exposed to
 a wider population.
--->

## Description
<!-- In detail, describe what your PR adds to PokeAlarm -->
This adds the ability to format an egg or raid alert to only occur when at a sponsored gym.  Sponsored gyms are identified by substrings provided via config file or command line such as [sprint,starbucks,level badge,phone skin]

Add the optional 'sponsored_raid' tag in the Egg and/or Raid sections of a format file.  For those alerts, the gym name will be checked to see if it includes the substrings provided earlier that indicate a sponsored gym.  It there is a match, the alert is allowed otherwise it is blocked.

## Type of Change
<!-- Place a single 'x' into the correct box, ex: [x] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (would cause existing functionality to change)

## Motivation and Context
<!---
 Why is this change required? What problem does it solve?
 If it fixes an open issue, please link to the issue here.
-->
Gives you the ability to alert on raids at sponsored gyms

## How Has This Been Tested?
<!---
 Please describe in detail how you tested your changes. Make sure to
 describe what tests you have performed, your testing environment,
 and if you have used this in a production setting.
-->
Tested on Ubuntu 16.04 in a production setting. 

## Screenshots (if appropriate):


## Wiki Update
<!--
 Does this feature require an update to the wiki? If so, please submit
 the required change to https://github.com/RocketMap/PokeAlarmWiki.
 If your feature requires a wiki update, you may submit it for review
 but it will not be accepted until the wiki update is complete.
--->
- [X] This change requires an update to the Wiki.
- [ ] This change does not require an update to the Wiki.


## Checklist
- [X] This change follows the code style of this project.
- [ ] This change only changes what is necessary to the feature.

tomballgithub commented:
Debugging an issue...

kvangent commented:
There are a few implementation issues with this, but I feel like it might be more useful as a general "contains" field. Instead of a 'sponsored_gyms' to true or false, you could specify the list of substrings to check for in the filter themselves. It would look something like this:
```
"contains": ["sprint", "starbucks"]
```
What are your thoughts on that?

tomballgithub commented:
My concern is I have 80+ filter files and it would be easier to change this in one place for sponsored_gyms than trying to touch it in all those files, even with a script.  For instance sprint gyms have had 3 names over the last few months in my area, 'sprint', ''get your level badge', and 'free team phone skin'

billyjbryant commented:
I agree with @kvangent's suggestion, having this be a filter to allow changing what text is contained within a gym name would be a better implementation.

tomballgithub commented:
I agree with that also if you want the flexibility of different trigger words for different managers, which is a problem not in search of a solution right now.  It works great if you have 1 or two alarms but it doesn't scale very well.

I was trying to create an easy way to handle a current need requested by many, which is a common set of trigger points easily managed across all managers.

kvangent commented:
So my general thought on this is that Managers are supposed to be replacements for standalone instances, but with this addition a user would require launching multiple instances. So I think we are unwilling to implement this in it's current form, but I'm going to leave it open for now for any users that are interested to check it out. 

I think you're real problem stems from having to have 80+ managers to satisfy your configuration needs, which I'm hoping will no longer be case in Patch 3.5.

tomballgithub commented:
I'll say this as clearly as possible - I cannot argue with your point 'that Managers are supposed to be replacements for standalone instances, but with this addition a user would require launching multiple instances [if they wanted to check for different gym names]'.  :-)  I will look at improving it

tomballgithub commented:
Requested change has been made and tested indirectly.  Will test under real world conditions in morning once raids start

billyjbryant commented:
@tomballgithub you can also use the webhook-tester in the `tools` folder to test your alerts.

tomballgithub commented:
Duh!  Didn't think of that.  Thank you...

kvangent commented on PokeAlarm/Filters.py:
```diff
@@ -140,7 +140,8 @@ def load_egg_section(settings):
         "min_level": int(settings.pop('min_level', 0) or 0),
-        "max_level": int(settings.pop('max_level', 10) or 10)
+        "max_level": int(settings.pop('max_level', 10) or 10),
+        "contains": settings.pop('gymname_contains', [])
```
We should probably do some check to ensure this is a list. If it's a string bad things will happen.

tomballgithub replied:
Can you recommend if/how I could do this with 'create_multi_filter' as I cannot figure that out?  Or should I just just create a separate function for that?

kvangent replied:
I don't think you should use create_mutli_filter. The safest bet is to just check if it's a list and raise a ValueException afterwards.

kvangent commented on PokeAlarm/Filters.py:
```diff
@@ -152,7 +153,8 @@ def load_egg_section(settings):
     raid = {
-        "enabled": bool(parse_boolean(settings.pop('enabled', None)) or False)
+        "enabled": bool(parse_boolean(settings.pop('enabled', None)) or False),
+        "contains": settings.pop('gymname_contains', [])
```
list check

kvangent commented on PokeAlarm/Manager.py:
```diff
@@ -847,6 +847,14 @@ def process_egg(self, egg):
+        gym_name = gym_info['name'].lower()
+
+        # Check if egg gym should be sponsored and is sponsored
+        if len(self.__egg_settings['contains']) > 0:
```
This should be after processed check because we don't want to keep getting notified about the same egg getting rejected over and over - I would suggest moving it to right before geofences.

kvangent commented on PokeAlarm/Manager.py:
```diff
@@ -887,14 +895,16 @@ def process_egg(self, egg):
         if self.__quiet is False:
-            log.info("Egg ({}) notification has been triggered!".format(gym_id))
+            if len(self.__egg_settings['contains']) > 0:
+                log.info("Sponsored Egg ({}) notification has been triggered!".format(gym_id))
```
No reason to differentiate these anymore.

kvangent commented on PokeAlarm/Manager.py:
```diff
@@ -847,6 +847,14 @@ def process_egg(self, egg):
 
         gym_id = egg['id']
+        gym_info = self.__cache.get_gym_info(gym_id)
+        gym_name = gym_info['name'].lower()
```
Instead of making this a variable, let's just use the actual `gym_info['name'].lower()` in the contains check. This will prevent anyone from accidentally using it and getting a name in lower case.

kvangent commented on PokeAlarm/Manager.py:
```diff
@@ -928,10 +938,18 @@ def process_raid(self, raid):
 
         gym_id = raid['id']
+        gym_info = self.__cache.get_gym_info(gym_id)
+        gym_name = gym_info['name'].lower()
```
same as above - (re)move to contains line

kvangent commented on PokeAlarm/Manager.py:
```diff
@@ -928,10 +938,18 @@ def process_raid(self, raid):
         raid_end = raid['raid_end']
 
+        # Check if raid gym should be sponsored and is sponsored
+        if len(self.__raid_settings['contains']) > 0:
```
same as above - move lower

kvangent commented on PokeAlarm/Manager.py:
```diff
@@ -997,14 +1015,16 @@ def process_raid(self, raid):
         if self.__quiet is False:
-            log.info("Raid ({}) notification has been triggered!".format(gym_id))
+            if len(self.__raid_settings['contains']) > 0:
+                log.info("Sponsored Raid ({}) notification has been triggered!".format(gym_id))
```
Same as above - no reason to call out sponsored

kvangent commented on filters.json.example:
```diff
@@ -426,11 +426,13 @@
     },
     "eggs":{
         "enabled":"True",
+        "gymname_contains":["sprint","level badge","phone skin","starbucks"],
```
I feel like this is better as `"contains"` but I'm interested to hearing why you decided on gymname_contains.

tomballgithub replied:
I used gymname_contains because it was crystal clear.  'Contains' by itself could mean other things such as a pokemon in the gym or a trainer in the gym or the type of pokemon for the raid or the street address or some other attribute.  I'd prefer something more succinct but I erred on the side of specificity in the code that was user-visible.  I think someone looking at a JSON that has this section filled out will know what 'contains' means but otherwise it requires looking at the wiki to be sure.

tomballgithub commented:
Made requested changes except I have a question on the list checking and if create_multi_filter can be used.  Made comment on your review.

kvangent commented:
Please remember you'll need to rebase to dev as well.

tomballgithub commented:
Added check to ensure lists are provided otherwise do sys.exit(1). 
Rebased...

kvangent commented:
Just to be clear, you have to rebase to the `dev` branch to pass the TravisCI check.

tomballgithub commented:
I just copied to a new branch SponsoredRaids4 and rebased to 'dev'.  Please check the results.  Looks like it is picking up a lot of other stuff.  I do not have the expertise to understand what is going on.  Any advice?

I did the following and fixed the conflicts after each iteration of the rebasing
```
git remote add upstream https://github.com/PokeAlarm/PokeAlarm.git
git fetch upstream
git rebase upstream/dev
git push -f origin SponsoredRaids4
```

billyjbryant commented:
@tomballgithub try using `rebase -i upstream/dev` and commenting out the commits that are not related to your PR.

kvangent commented:
Additionally, when you rebase a branch you are rewriting it's history. You'll have to use `git push <remote_name> <branch_name> -f` to 'force' a rewrite. Or, if you pushed it to a new branch, you'll have to edit your PR to point to the new branch.

tomballgithub commented:
git remote add upstream https://github.com/PokeAlarm/PokeAlarm.git
git fetch upstream
git -i rebase upstream/dev
git push -f origin SponsoredRaids

Tried again and then reset it back the way it was.  I get 41 files changed and most of it has nothing to do with my PR.  I saw the comment about doing an -i rebase, which I did, but not of the commits listed needed to be commented out, they were ALL related to my PR.

tomballgithub commented:
I believe 'origin' is the master branch

tomballgithub commented:
All rebased correctly.

FYI, I had someone ask me coincidentally if there was a way they could make a manager that only triggered on raids with the word 'park' in them.  This method will support that as you guys recommended.

tomballgithub commented:
And, the way this is written, with the filter being optional, no existing filter files need to be changed.  So although I have 100+ sets of manager files only 1 is for sponsored raids and so that is all that needs editing

kvangent commented:
@tomballgithub the travisci check is showing some pep8 coding style violations. Could you please fix those so we can pull this?

You can see them by clicking 'Details' next to the travis status check.

You can run flake8 on your own machine by installing via pip, and then running `flake8 --show-statistics --show-source`

tomballgithub commented:
All fixed...

kvangent commented on PokeAlarm/Manager.py:
```diff
@@ -981,8 +981,11 @@ def process_egg(self, egg):
-            if not any(x in gym_info['name'].lower() for x in self.__egg_settings['contains']):
-                log.info("Egg {} ignored: gym name did not match contains filter.".format(gym_info['name'].lower()))
+            if not any(x in gym_info['name'].lower()
+                       for x in self.__egg_settings['contains']):
```
Have you tested this? It looks like your indentation might be off, and only the first gym_name contains will be checked

tomballgithub commented:
I tested and verified that the checking of the gym name with the formatting change works as desired

I did find that spreading the log.info format string across multiple lines with + did not work and all {} were printed as {} with no substitution.  Removing the + and letting python automatically concatenate adjacent strings is acceptable and works.

Changed to gym_id in the log message instead of the gym name for consistency

tomballgithub commented:
Someone who pulled my fork/branch directly says this is not triggering as expected.  I will investigate including adding log.debug information for future

blakewenloe commented:
It would be amazing if I could check against an array of gym names that I can input/change as needed. That would only require one filter file.

tomballgithub commented:
I have extensive logs showing that this feature works.  I added some optional debug logging.  And fixed Travis.

kvangent commented:
Can you confirm this is up to date with the dev branch? It says it's out of date and failing Tavis.

tomballgithub commented:
Yes, there was a travis problem of two new debug lines being too long.  Fixed and passing...

tomballgithub commented:
And I confirmed the person who was having issues with my branch is using Monocle and was not receiving the gym_name information via Monocle which is why they were having issues

billyjbryant commented on filters.json.example:
```diff
@@ -426,11 +426,13 @@
     },
     "eggs":{
         "enabled":"True",
+        "gymname_contains":["sprint","level badge","phone skin","starbucks"],
```
Filter examples need to be updated with the change in verbiage and to match the actual checks.

tomballgithub replied:
'gymname_contains' is the verbiage.  can you clarify what you are asking?

billyjbryant commented on filters.json.example:
```diff
@@ -426,11 +426,13 @@
     },
     "raids":{
         "enabled":"True",
+        "gymname_contains":["sprint","level badge","phone skin","starbucks"],
```
Same as above.

tomballgithub replied:
'gymname_contains' is the verbiage.  can you clarify what you are asking?

billyjbryant replied:
Nevermind, I misread. You are good.
